flipt-engine-ffi/src/http.rs

```diff
@@ -148,209 +228,376 @@ impl Parser for HTTPParser {
-            // check if we have a 304 response
             reqwest::StatusCode::NOT_MODIFIED => Ok(None),
+
             reqwest::StatusCode::OK => {
                 // check if we have a new etag
                 if let Some(etag) = response.headers().get(reqwest::header::ETAG) {
                     self.etag = Some(etag.to_str().unwrap().to_string());
                 }
 
-                let response_text = match response.text() {
-                    Ok(t) => t,
-                    Err(e) => {
-                        return Err(Error::Server(format!("failed to get response body: {}", e)))
-                    }
-                };
-
-                match serde_json::from_str(&response_text) {
-                    Ok(doc) => Ok(Some(doc)),
-                    Err(e) => Err(Error::InvalidJSON(e.to_string())),
-                }
+                Ok(Some(response))
             }
+
             _ => Err(Error::Server(format!(
                 "unexpected http response: {} {}",
                 response.status(),
-                response.text().unwrap_or("".to_string())
+                response.text().await.unwrap_or("".to_string())
             ))),
         }
     }
+
+    async fn fetch_stream(&mut self) -> Result<Option<Response>, Error> {
+        let url = format!(
+            "{}/internal/v1/evaluation/snapshots?[]namespaces={}",
+            self.base_url, self.namespace
+        );
+
+        self.http_client
+            .get(url)
+            .headers(self.build_headers())
+            .send()
+            .await
+            .map_err(|e| Error::Server(format!("failed to make request: {}", e)))?
+            .error_for_status()
+            .map_err(|e| Error::Server(format!("response: {}", e)))
+            .map(Some)
+    }
+
+    async fn handle_polling(
+        &mut self,
+        sender: &mpsc::Sender<Result<source::Document, Error>>,
+    ) -> Result<(), Error> {
+        let result = match self.fetch().await {
+            Ok(Some(resp)) => match resp.text().await {
+                Ok(json) => match serde_json::from_str(&json) {
+                    Ok(doc) => Ok(doc),
+                    Err(e) => Err(Error::InvalidJSON(format!(
+                        "failed to parse response body: {}",
+                        e
+                    ))),
+                },
+                Err(e) => Err(Error::Server(format!(
+                    "failed to read response body: {}",
+                    e
+                ))),
+            },
+            Ok(None) => return Ok(()),
+            Err(e) => Err(e),
+        };
+
+        sender
+            .send(result)
+            .await
+            .map_err(|_| Error::Server("failed to send result".into()))
+    }
+
+    async fn handle_streaming(
+        &mut self,
+        sender: &mpsc::Sender<Result<source::Document, Error>>,
+    ) -> Result<(), Error> {
+        let stream_result = self.fetch_stream().await;
+
+        match stream_result {
```

Comment:
Maybe it's too late. Could serde_json::StreamDeserializer be used here?

Reply:
maybe, this works however. do you know if `serde_json::StreamDeserializer` would provide a benefit?

Reply:
I am worry about bytes and `\n` on utf-8 string. But it could be done with refactoring.